core/js/typespecs/app_types.ts

```diff
@@ -173,39 +173,84 @@ export const datasetSchema = z
-      ),
-    project_id: promptBaseSchema.shape.project_id,
-    log_id: z
-      .literal("p")
-      .describe(
-        "A literal 'p' which identifies the object as a project prompt",
-      ),
-    org_id: organizationSchema.shape.id,
-    name: promptBaseSchema.shape.name,
-    slug: z.string().describe("Unique identifier for the prompt"),
-    description: promptBaseSchema.shape.description,
-    created: promptBaseSchema.shape.created,
-    prompt_data: promptDataSchema
-      .nullish()
-      .describe("The prompt, model, and its parameters"),
-    tags: z
-      .array(z.string())
-      .nullish()
-      .describe("A list of tags for the prompt"),
-    metadata: promptBaseSchema.shape.metadata,
-  })
-  .openapi("Prompt");
+const promptSchemaObject = z.strictObject({
+  id: promptBaseSchema.shape.id,
+  // This has to be copy/pasted because zod blows up when there are circular dependencies
+  _xact_id: z
+    .string()
+    .describe(
+      `The transaction id of an event is unique to the network operation that processed the event insertion. Transaction ids are monotonically increasing over time and can be used to retrieve a versioned snapshot of the prompt (see the \`version\` parameter)`,
+    ),
+  project_id: promptBaseSchema.shape.project_id,
+  log_id: z
+    .literal("p")
+    .describe("A literal 'p' which identifies the object as a project prompt"),
+  org_id: organizationSchema.shape.id,
+  name: promptBaseSchema.shape.name,
+  slug: z.string().describe("Unique identifier for the prompt"),
+  description: promptBaseSchema.shape.description,
+  created: promptBaseSchema.shape.created,
+  prompt_data: promptDataSchema
+    .nullish()
+    .describe("The prompt, model, and its parameters"),
+  tags: z.array(z.string()).nullish().describe("A list of tags for the prompt"),
+  metadata: promptBaseSchema.shape.metadata,
+});
+
+export const promptSchema = promptSchemaObject.openapi("Prompt");
 export type Prompt = z.infer<typeof promptSchema>;
 
+export const codeBundleSchema = z.strictObject({
+  runtime_context: z.strictObject({
+    runtime: z.enum(validRuntimes),
+    version: z.string(),
+  }),
+  // This should be a union, once we support code living in different places
+  // Other options should be:
+  //  - a "handler" function that has some signature [does AWS lambda assume it's always called "handler"?]
+  location: z.strictObject({
+    type: z.literal("experiment"),
+    eval_name: z.string(),
+    position: z.union([
+      z.literal("task"),
+      z.strictObject({ score: z.number() }),
+    ]),
+  }),
+  bundle_id: z.string(),
+});
+export type CodeBundle = z.infer<typeof codeBundleSchema>;
+
+export const functionDataSchema = z.union([
+  z.strictObject({
+    type: z.literal("prompt"),
```

Comment:
Maybe a comment stating that for backwards-compatibility reasons, the prompt data is hoisted outside this object.